networkit/cpp/embedding/BiasedRandomWalk.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * BiasedRandomWalk.cpp
+ *
+ *  Created on: 03.07.2020
+ *      Author: Klaus Ahrens  <ahrens@informatik.hu-berlin.de>
+ *
+ *  adapted and reimplemented from node2vec
+ *  part of snap [https://github.com/snap-stanford/snap]
+ *  Copyright (c) 2007-2019, Jure Leskovec (under BSD license)
+ *
+ *  see [https://arxiv.org/pdf/1607.00653v1.pdf]
+ *
+ */
+
+// networkit-format
+
+#include <algorithm>
+#include <utility>
+
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/graph/Graph.hpp>
+
+#include "AliasSampler.hpp"
+#include "BiasedRandomWalk.hpp"
+
+namespace NetworKit {
+namespace Embedding {
+
+using Walk = BiasedRandomWalk::Walk;
+using AllWalks = BiasedRandomWalk::AllWalks;
+
+BiasedRandomWalk::BiasedRandomWalk(const Graph *graph)
+    : graph(graph), graphData(std::make_unique<GraphData>(graph->numberOfNodes())) {
+    auto nn = graph->numberOfNodes();
+    index2node.resize(nn);
+    // pre-allocate unordered maps:
+    graph->forNodes([&](node v) {
+        auto degv = graph->degreeOut(v);
+        for (auto n : graph->neighborRange(v)) {
+            // init index2node:
+            index2node[v].push_back(n);
+            graphData->data[v][n] = AliasSampler(degv);
+        }
+    });
+}
+
+void BiasedRandomWalk::preprocessNode(node t, double paramP, double paramQ) {
+
+    // for node t
+    NeighborSet tNbrs; // Neighbors of t
+    for (auto i : graph->neighborRange(t)) {
+        tNbrs.insert(i);
+    }
+    // for each t-neighbor v:
+    for (auto v : graph->neighborRange(t)) {
+        double pSum = 0;
+        std::vector<float> pTable; // Probability distribution table
+        // for each v-neighbor
+        graph->forNeighborsOf(v, [&](node x, edgeweight weight) {
+            if (x == t) {
+                pTable.push_back(weight / paramP);
+                pSum += weight / paramP;
+            } else if (tNbrs.count(x)) {
+                pTable.push_back(weight);
+                pSum += weight;
+            } else {
+                pTable.push_back(weight / paramQ);
+                pSum += weight / paramQ;
+            }
+        });
+        // Normalizing table
+        float pfSum = (float)pSum;
+        std::for_each(pTable.begin(), pTable.end(), [pfSum](float &p) { p /= pfSum; });
+
+        graphData->data[v][t].unigram(pTable);
```

Comment:
This looks dangerous to me, as `graphData->data[v]` is an `std::unordered_map`, and this code is called in parallel loop over `t`. Or, in other words, why cannot two threads write to the unordered map (which is not thread-safe) of the same `v`?

Reply:
Each `graphData->data[v] ` keeps a map of nextStep-probabilities depending on the previous step (coming from t). So while processing these in parallel over all nodes, every slot is filled exactly once. These maps are preallocated in the number of neighbors, so that the memory behind the maps does not change - only the values (AliasSampler) are changed from 0 to the calculated probs. That's why I think, there is no threading issue.

Reply:
Ah, I see. Maybe this could be made clearer by explicitly using a method (like `find()`), that does not potentially trigger a modification of the `unordered_map`?